community/blog/2023-01 Role-based authentication in Azure API Management made easy in Arcus Scripting v0.8/README.md

```diff
@@ -0,0 +1,63 @@
+
+An example of such a policy is shown here:
+
+```xml
+<validate-jwt header-name="Authorization" failed-validation-httpcode="401" failed-validation-error-message="Unauthorized. Access token is missing or invalid.">
+    <openid-config url="{{openid-config-url}}" />
+    <issuers>
+        <issuer>{{issuer}}</issuer>
+    </issuers>
+    <required-claims>
+        <claim name="roles" match="any">
+            <value>DummyRole</value>
+        </claim>
+    </required-claims>
+</validate-jwt>
+```
+
+Management of service principal roles to certain Azure Directory Applications is rather tedious. One has to look up the role assignments of an Azure Active Directory application to find out if the service principal has the correct access. Moreover, in certain scenarios, one has to wait a couple of seconds before a role assignment is available for use. All this adds to the problem of managing a service principal for validating JWTs in Azure API Management.
+
+## List, add, and remove role assignments for a service principal
+In a single Arcus Scripting release, we have fully fixed the problem of role assignments for a service principal. In a new `Arcus.Scripting.ActiveDirectory` PowerShell module, we have created three functions that let you list, add, and remove role assignments to let a service principal access an Azure Active Directory.
+
+```powershell
+PS> List-AzADAppRoleAssignments `
+  -ClientId 'b885c208-6067-44bd-aba9-4010c62b7d85' `
+  -RolesAssignedToClientId '6ea09bbd-c21c-460c-b58a-f4a720f51826'
+#Found role 'FirstRole' on Active Directory Application 'main-application'
+#Role 'FirstRole' is assigned to the Active Directory Application 'client-application-one' with id '6ea09bbd-c21c-460c-b58a-f4a720f51826'
+#Found role 'SecondRole' on Active Directory Application 'main-application'
+#Role 'SecondRole' is assigned to the Active Directory Application 'client-application-one' with id '6ea09bbd-c21c-460c-b58a-f4a720f51826'
+
+PS> Add-AzADAppRoleAssignment `
+  -ClientId "b885c208-6067-44bd-aba9-4010c62b7d85" `
+  -Role "DummyRole" `
+  -AssignRoleToClientId "6ea09bbd-c21c-460c-b58a-f4a720f51826"
+#Active Directory Application 'main-application' does not contain the role 'DummyRole', adding the role
+#Added Role 'DummyRole' to Active Directory Application 'main-application'
+#Role Assignment for the role 'DummyRole' added to the Active Directory Application 'client-application-one'
+
+PS> Remove-AzADAppRoleAssignment `
+  -ClientId "b885c208-6067-44bd-aba9-4010c62b7d85" `
+  -Role "DummyRole" `
+  -RemoveRoleFromClientId "6ea09bbd-c21c-460c-b58a-f4a720f51826" `
+#Role assignment for 'DummyRole' has been removed from Active Directory Application 'client-application-one'
+```
```

Comment:
Playing the devil's advocate here, but why would I use these Arcus scripts instead of the native Azure CLI commands:
https://learn.microsoft.com/en-us/cli/azure/role/assignment?view=azure-cli-latest#az-role-assignment-create
 ?

Reply:
@pim-simons will be better suited to answer this, but as I understand it, it is about how our scripts are more fail-safe as they they into account existing roles and has a 'return until it's available' feature so that after the script has run, you know that the role assignment is ready to use.

Reply:
Correct, and it combines several tasks into one script (creating the role and the role assignment).